Fix POST requests

justincy opened the pull request:
Here are failing test cases for #120. The tests seem to be exhibiting a different, though perhaps related, problem than what I was seeing. I'll have to debug further to figure out what's going wrong. Any ideas?

adri commented:
Thank you very much for you test cases!

adri commented:
I looked into it and it seems to be an issue with the virtual file system. When I change the test `setUp` method to use `\VCR\VCR::configure()->setCassettePath('/tmp');` the tests run fine.

adri commented:
From the recorded fixture file:

```
    request:
        method: POST
        url: 'http://httpbin.org/post'
        headers:
            Host: httpbin.org
            Expect: null
            Content-Type: null
            Accept-Encoding: null
            User-Agent: 'Guzzle/5.1.0 curl/7.38.0 PHP/5.5.20'
            Accept: null
        body: foo=bar
```

justincy commented:
@adri That change doesn't fix the tests for me. Maybe it's a problem with a specific PHP + curl version. I'm using PHP 5.5.9 with curl version 7.35.0.

justincy commented:
@adri Removing vfs fixes Guzzle 5 but not 4 or 3.

adri commented:
Thanks for the info. I noticed that the error message is different for [Guzzle 3](https://travis-ci.org/php-vcr/php-vcr/jobs/66870715).

adri commented:
Seems like in Guzzle version 3 there is some CURL option set which leads to the the following error [HttpClient](https://github.com/php-vcr/php-vcr/blob/master/src/VCR/Util/HttpClient.php#L34
), `curl_error($ch)` returns:
`Recv failure: Connection reset by peer`

adri commented:
It seems to be the curl option `CURLOPT_UPLOAD` which is set in [guzzle 3 `CurlHandle:factory()`](https://github.com/guzzle/guzzle3/blob/master/src/Guzzle/Http/Curl/CurlHandle.php#L127). When I uncomment `$curlOptions[CURLOPT_UPLOAD] = true;` the request works.

aaa2000 commented:
Using the PHPUnit annotation @runTestsInSeparateProcesses fixes also Guzzle5

justincy commented:
@adri When I comment out that line the request succeeds but the test fails because the POST body isn't set.

adri commented:
Yes indeed, I haven't found out the difference between the intercepted and the direct request yet.   I looked at the HttpClient what curl options are provided. Something there must set the body and make the request not "timeout" (error posted previously). Maybe it is related to the CURLOPT_READFUNCTION, I have to check.

> On 17.06.2015, at 16:38, Justin <notifications@github.com> wrote:
> 
> @adri When I comment out that line the request succeeds but the test fails because the POST body isn't set.
> 
> —
> Reply to this email directly or view it on GitHub.
>

justincy commented:
It looks like I may be using version 3 wrong. It's treating by post body string as a stream. It also expects the `Content-Length` header to contain the body size for `CURLOPT_INFILESIZE`. That behavior is undocumented so you'd think they would calculate that on their but it doesn't appear to be doing that.

justincy commented:
Though the direct request is still working, so maybe cURL is just really smart.

justincy commented:
I can get the requests to pass and almost all pass if I comment out line 28 of `HttpClient` in VCR:

```php
curl_setopt_array($ch, $request->getCurlOptions());
```

So Guzzle must be doing something unexpected with the curl opts. Perhaps it's related to streaming, since VCR intercepts streams and sends them all as strings via `CURLOPT_POSTFIELDS`.

justincy commented:
That's the `CURLOPT_UPLOAD` that you were talking about. Commenting that out works for me now. I think that wasn't working earlier for me because it was detecting form data and handling it differently and so the response from httpbin was different than expected. I'm testing with JSON now to avoid that. I think I'm close to fixing VCR for guzzle3.

justincy commented:
I think I've fixed them all.

To fix Guzzle 3 I had to flat out ignore `CURLOPT_UPLOAD`. That could be a bad idea.

To fix Guzzle 4 I first changed when the `CURLOPT_READFUNCTION, CURLOPT_INFILESIZE` assertion was being done so that `CURLOPT_INFILESIZE` didn't have to be set before `CURLOPT_READFUNCTION`. That didn't actually fix Guzzle 4 though. The problem there was that they were setting `CURLOPT_READFUNCTION` by default and then later setting `CURLOPT_POSTFIELDS` for small strings without bothering to set `CURLOPT_INFILESIZE`. So I changed it to ignore `CURLOPT_READFUNCTION` when the body has already been set via `CURLOPT_POSTFIELDS`.

Looks like the tests failed for Guzzle 4. I don't know why. They pass for me.

justincy commented:
Now we see green!

adri commented:
Great work! I'll review.

adri commented on src/VCR/Util/CurlHelper.php:
```diff
@@ -175,24 +175,39 @@ public static function setCurlOptionOnRequest(Request $request, $option, $value,
             case CURLOPT_HEADER:
             case CURLOPT_WRITEFUNCTION:
             case CURLOPT_HEADERFUNCTION:
+            case CURLOPT_UPLOAD:
```
As far as I see [`CURLOPT_UPLOAD`](http://curl.haxx.se/libcurl/c/CURLOPT_UPLOAD.html) should be used when uploading a file in conjunction with a read function. So it should be still possible to upload files using the `@` notation when we ignore the option.

adri commented on src/VCR/Util/CurlHelper.php:
```diff
@@ -175,24 +175,39 @@ public static function setCurlOptionOnRequest(Request $request, $option, $value,
+            return;
+        }
+        
+        // Guzzle 4 sometimes sets the post body in CURLOPT_POSTFIELDS even if
```
When did you come across this case? As far as I see it's either an upload (with a read function) or a string. See   [CurlFactory](https://github.com/guzzle/guzzle/blob/ddbbba92efbf3d843baf467dd699b4647105c64a/src/Handler/CurlFactory.php#L224).

justincy replied:
You linked to the current version which is Guzzle 5. The case I was handling occurs in [Guzzle 4](https://github.com/guzzle/guzzle/blob/4.2.3/src/Adapter/Curl/CurlFactory.php). In `getDefaultOptions` you can see that `CURLOPT_READFUNCTION` is set by default while in `applyBody` you can have `CURLOPT_POSTFIELDS` set instead of `CURLOPT_UPLOAD` and `CURLOPT_INFILESIZE`.

adri commented on src/VCR/Util/CurlHelper.php:
```diff
@@ -175,24 +175,39 @@ public static function setCurlOptionOnRequest(Request $request, $option, $value,
                 // Ignore header, file and writer functions.
                 // These options are stored and will be handled later in handleOutput().
                 break;
-            case CURLOPT_READFUNCTION:
```
Removing handling the curl option here was probably necessary because it is not clear if `CURLOPT_POSTFIELDS` have been set yet, right? As far as I understand this came from a special case when using Guzzle 4. Do you think it is possible to leave it here and check for the curl option because Guzzle 4 sets it before the `CURLOPT_READFUNCTION` option?

justincy replied:
I don't think we can move it. As you see in Guzzle 4 that I just linked to, `CURLOPT_READFUNCTION` is set before `CURLOPT_POSTFIELDS`. Plus this also avoids the problem of needing to set `CURLOPT_INFILESIZE` before `CURLOPT_READFUNCTION`.

adri commented on tests/integration/guzzle/3/ExampleHttpClient.php:
```diff
@@ -24,6 +24,25 @@ public function get($url)
+        
+        try {
+            $request = $client->post($url);
+            $request->setBody($body, 'application/json');
```
Just for me to know, is the 'application/json' parameter changing anything how Guzzle sends the request? (Except from a `Content-Type` header?

justincy replied:
No it doesn't change anything other than setting the `Content-Type` header.

adri commented on tests/integration/guzzle/4/ExampleHttpClient.php:
```diff
@@ -22,4 +22,20 @@ public function get($url)
+        $client = new Client();
+        
+        try {
+            $response = $client->post($url, ['body' => $body]);
```
PHP-VCR is still supporting PHP 5.3, please use `array()`.

adri commented on tests/integration/guzzle/5/ExampleHttpClient.php:
```diff
@@ -22,4 +22,20 @@ public function get($url)
+        $client = new Client();
+        
+        try {
+            $response = $client->post($url, ['body' => $body]);
```
PHP-VCR is still supporting PHP 5.3, please use `array()`.

adri commented on tests/integration/guzzle/test/VCR/Example/ExampleHttpClientTest.php:
```diff
@@ -20,19 +24,18 @@ class ExampleHttpClientTest extends \PHPUnit_Framework_TestCase
-        // Configure virtual filesystem.
-        vfsStream::setup('testDir');
-        \VCR\VCR::configure()->setCassettePath(vfsStream::url('testDir'));
+        \VCR\VCR::configure()->setCassettePath('/tmp');
```
We should still use vfsStream. In theory this shouldn't be a problem to use. Writing to the file system is not handy for repeatable tests since the fixture file then needs to be deleted.

justincy commented:
@adri I stopped using vfsStream because it fails for some reason when recording POST requests. I put it back to show the errors. How do you suggest we handle this?

aaa2000 commented:
There are issues that can not be fixed in vfsStream https://github.com/mikey179/vfsStream/wiki/Known-Issues

> If you use an object with a destructor that opens a vfsStream URL, vfsStream can become unavailable before the open occurs, causing it to fail with a WARNING. E.g. in PHPUnit, make sure to manually destroy the object in the test or in tearDown().

I think the problem is that VCR use a singleton, the same instance of videorecorder and same cassette is used during all the tests.

So, I see two way:
- the @runTestsInSeparateProcesses annotation on top of the class https://phpunit.de/manual/current/en/appendixes.annotations.html#appendixes.annotations.runTestsInSeparateProcesses
```
/**
 * @runTestsInSeparateProcesses
 * @preserveGlobalState disabled
 */
```
- or not using \VCR\VCR class in ExampleHttpClientTest and bootstrap.php

```
   public function setUp()
    {
        // Configure virtual filesystem.
        vfsStream::setup('testDir');
        $this->configuration = new \VCR\Configuration();
        $this->configuration->setCassettePath(vfsStream::url('testDir'));
        $this->httpClient = new \VCR\Util\HttpClient();
        $this->factory = \VCR\VCRFactory::getInstance($this->configuration);
        $this->videorecorder = new \VCR\Videorecorder($this->configuration, $this->httpClient, $this->factory);
    }
...
    protected function requestGETIntercepted()
    {
        $this->videorecorder->turnOn();
        $this->videorecorder->insertCassette('test-cassette.yml');
        $info = $this->requestGET();
        $this->videorecorder->turnOff();

        return $info;
    }

    protected function requestPOSTIntercepted()
    {
        $this->videorecorder->turnOn();
        $this->videorecorder->insertCassette('test-cassette.yml');
        $info = $this->requestPOST();
        $this->videorecorder->turnOff();

        return $info;
    }
```

adri commented:
Great analysis @aaa2000 thank you. 

When [turning off VCR](https://github.com/php-vcr/php-vcr/blob/master/src/VCR/Videorecorder.php#L135) the cassette instance is actually unset. But you are right the storage will then be retrieved from the Factory which is the same instance for the same file.

There might be another way. That assertion for a writable file leads to quite some problems, see for example #103. Maybe on some systems `is_writeable` returns `false` although writing would be possible.

I propose to remove the assertion. Directly after asserting for writeablility the file is written to. So when it's not writeable that code will just fail anyway. That means that in [Yaml](https://github.com/php-vcr/php-vcr/blob/master/src/VCR/Storage/Yaml.php#L48) and in [Json](https://github.com/php-vcr/php-vcr/blob/master/src/VCR/Storage/Json.php#L20) we could remove the assertion and fix the tests. @justincy What do you think, want to give it a try? My local test got green then. I guess then you can confirm, we can merge.

justincy commented:
@adri Fixed.

adri commented:
Awesome, well done.
